### ISSUE TITLE
Generate referenced OpenApi schema

### ISSUE THREAD

**Issue by krzemin** [julienrf/endpoints#108]
Currently `endpoints` has a simple way of generating OpenApi request/response schema: it always inline produced Json schema for referenced request/response entity type.

While it's simple and works really great, this approach has few drawbacks:
- generated documentation is redundant and oversized, especially when a single large type is referred from several endpoints
- it undermines usage of code generation tools that takes this `endpoints`-generated openapi spec as a source; there are actually [quite a lot of them](https://github.com/OpenAPITools/openapi-generator#overview)

OpenApi [allows](https://swagger.io/docs/specification/using-ref/) to store the schema in `/components/schema/...` section and using local paths like `"$ref": "#/components/schemas/Foo` to refer the schema from particular endpoint definitions.

It would be really nice if `endpoints` would (optionally or by default) support this way of generating OpenApi documentation. I'm not yet sure how difficult to implement it would be, but I can definitely offer my help.

**Comment by Fristi** [comment]
I did it in my extension project of `endpoints`.

1. Per endpoint I interpret it as `SwaggerOperation` (https://github.com/Fristi/endpoints-proto/blob/master/swagger/src/main/scala/itinere/swagger/Types.scala#L91)
2. You supply all your `SwaggerOperation`'s to this method https://github.com/Fristi/endpoints-proto/blob/master/swagger/src/main/scala/itinere/swagger/SwaggerGen.scala#L19
3. The `toReferenceTree` will rewrite the full blown schema AST's to a referenced version. Equiped with a `Writer` Monad, it will build up a `SchemaManifest`, this is a `Map[String, Schema]`. The keys are the names of references (which are prefixed by `"#/components/schemas/`) and `Schema` are the rewritten `Schema` definitions.

Hope this helps :-)

**Comment by krzemin** [comment]
Thanks, that looks very promising - I was able to run your example and it works indeed exactly as I would like to see it!

<img width="1413" alt="zrzut ekranu 2018-08-20 o 17 02 26" src="https://user-images.githubusercontent.com/2477704/44348512-ecd37500-a49a-11e8-9e98-26e4375fa274.png">

I would definitely want to see this ported into upstream as integration with 3rd party, non-released library is a bit problematic. And I think every `endpoints` user deserve for such quality of generated documentation.

@Fristi did you think about porting your work to the library? Or can you shed a light on any potential issues that might cause problems when doing such port?

**Comment by Fristi** [comment]
Yes I have thinking about porting my work. Some parts are a bit different in my project.

- How the JSON schema is build
- The algebra has some (subtle) differences
- `endpoints` does openapi, while I think this my implementation is also close to open api, everything is named _swagger_. So there could be subtle changes needed to port it to a open api specification

I think it is doable. I was working on a separate library for defining JSON schemas.. but I got derailed a bit :-) However I have limited bandwidth to work on these kinds of things nowadays, so if you would like to give it a stab go a head :-)
